### PR TITLE
chore(Platform): Integrate example-app

### PR DESCRIPTION
**Pull request by MikeRyanDev**
Migrates the example-app into the monorepo. This makes the example-app a viable target for complete ngrx integration tests. Right now the app _mostly_ works. Things that block this PR:

- [ ] @ngrx/router-store breaks badly when instrumented with the Redux Devtools Extension. This is because the router's state is not completely serializable cc @vsavkin 
- [x] @angular/cli has a hard dependency on TypeScript 2.2 whereas we have a hard min dependency on TypeScript 2.3. Run `yarn run clean:ts` before trying to start the development server.
- [x] Both @ngrx/store and @ngrx/store-devtools now have mixed eager providers making it hard to debug `undefined` injector errors, see https://github.com/angular/angular/issues/15501
- [x] Example app unit tests use @ngrx/effects/testing which I do not plan to carry over for @ngrx/effects v4. These should instead be rewritten to use marble diagram tests cc @krjordan

### CODE REVIEW

File: example-app/app/app.module.ts [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * the store as the single source of truth for the router's state.
+     */
+    StoreRouterConnectingModule,
+
+    /**
+     * Store devtools instrument the store retaining past versions of state
+     * and recalculating new states. This enables powerful time-travel
+     * debugging.
+     *
+     * To use the debugger, install the Redux Devtools extension for either
```
Comment (Matmo10) [inline example-app/app/app.module.ts]: Is the `instrumentOnlyWithExtension()` functionality baked into `instrument()` now?

Comment (MikeRyanDev) [inline example-app/app/app.module.ts]: Yes, there is now just one API for instrumentation regardless of strategy.

**Comment by vsavkin** [comment]
@MikeRyanDev the router state is not serializable because it has a cycle. Breaking the cycle would be a breaking change in the router and will have some negative implications on dev ergonomics. Is there a way to insert a function serializing the state?

**Comment by MikeRyanDev** [comment]
@vsavkin Yes, we can write a custom serializer/parser for the dev tools that handles router state in a special way. The interface for these will have to be something like:

```ts
function serialize<T>(state: RouterState): T;

function revive<T>(state: T): RouterState;
```

where `T` is some shape that can be passed to `JSON.stringify()`.

**Comment by vsavkin** [comment]
@MikeRyanDev I can write one if needed. 

Why does it have to be serializable at all?

**Comment by Matmo10** [comment]
@MikeRyanDev Just FYI (you're probably already aware), `deepFreeze`ing the router state also causes errors.

File: modules/store/spec/store.spec.ts [hunk 2]
```diff
@@ -187,7 +187,6 @@ describe('ngRx Store', () => {
 
       dispatcher.ngOnDestroy();
 
-      expect(storeSubscription.closed).toBe(true);
```
Comment (brandonroberts) [inline modules/store/spec/store.spec.ts]: @MikeRyanDev this test was failing as the `storeSubscription` is no longer closed. It didn't show up until I upgraded RxJS. Is this a change in behavior?

Comment (MikeRyanDev) [inline modules/store/spec/store.spec.ts]: Yeah, I'm seeing the same thing. I think this behavior changed in newer versions of rxjs.

**Comment by brandonroberts** [comment]
@MikeRyanDev this branch is up-to-date and ready for another review

File: example-app/app/books/books.module.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@NgModule({
+  imports: [
+    BooksModule,
+    /**
+     * Effects.forFeature is used to regiser effects
+     * from feature modules. Effects can be loaded
+     * eagerly or lazily and will be started immediately.
+     * 
+     * All Effects will only be instantiated regardless of
```
Comment (MikeRyanDev) [inline example-app/app/books/books.module.ts]: "instantiated once"

File: example-app/app/books/effects/book.spec.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        { provide: SEARCH_DEBOUNCE, useValue: 30 },
+      ]
+    });
+
+    effects = TestBed.get(BookEffects);
+    googleBooksService = TestBed.get(GoogleBooksService);
+    actions$ = TestBed.get(Actions);
+  });
+
+  describe('search$', () => {
```
Comment (MikeRyanDev) [inline example-app/app/books/effects/book.spec.ts]: I hope you are prepared to start supporting marble testing questions... 😄

File: example-app/app/books/effects/book.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      return this.googleBooks.searchBooks(query)
+        .takeUntil(nextSearch$)
+        .map((books: Book[]) => new book.SearchCompleteAction(books))
+        .catch(() => of(new book.SearchCompleteAction([])));
+    });
+
+    constructor(
+      private actions$: Actions,
+      private googleBooks: GoogleBooksService,
+      @Optional() @Inject(SEARCH_SCHEDULER) private scheduler: Scheduler,
```
Comment (MikeRyanDev) [inline example-app/app/books/effects/book.ts]: I would maybe add a comment about why these are necessary for testability purposes.

File: example-app/app/reducers/index.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  return function(state: State, action: any) {
+    console.log('state', state);
+    console.log('action', action);
+
+    return reducer(state, action);
+  }
+}
+
+/**
+ * The compose function is one of our most handy tools. In basic terms, you give
```
Comment (MikeRyanDev) [inline example-app/app/reducers/index.ts]: I don't think many of these comments apply anymore. Maybe add a comment about what `createSelector` does?

File: example-app/app/books/effects/book.spec.ts (hunk 4 above)
Comment (brandonroberts) [inline example-app/app/books/effects/book.spec.ts]: 😎

File: example-app/app/reducers/index.ts (hunk 6 above)
Comment (brandonroberts) [inline example-app/app/reducers/index.ts]: Do we not need the comments for showing how the `reducerFactory` works?

File: example-app/app/books/effects/book.ts (hunk 5 above)
Comment (brandonroberts) [inline example-app/app/books/effects/book.ts]: Fixed

File: example-app/app/books/books.module.ts (hunk 3 above)
Comment (brandonroberts) [inline example-app/app/books/books.module.ts]: Fixed

**Comment by sharikovvladislav** [comment]
Thank you!

**Comment by sharikovvladislav** [comment]
Is it planned that example app does not work after merging this?

**Comment by sharikovvladislav** [comment]
I get this on `ng serve` right after `yarn install`:

```
WARNING in ./modules/store-devtools/src/devtools.ts
122:523-542 "export 'StoreDevtoolsConfig' was not found in './config'

WARNING in ./modules/store-devtools/src/devtools.ts
122:562-581 "export 'StoreDevtoolsConfig' was not found in './config'

ERROR in Error encountered resolving symbol values statically. Function calls are not supported. Consider replacing the function or lambda with a reference to an exported function (position 47:10 in the original .ts file), resolving symbol compose in /Users/svlad/dev/platform/modules/store/src/utils.ts, resolving symbol developmentReducerFactory in /Users/svlad/dev/platform/example-app/app/reducers/index.ts, resolving symbol AppModule in /Users/svlad/dev/platform/example-app/app/app.module.ts, resolving symbol AppModule in /Users/svlad/dev/platform/example-app/app/app.module.ts, resolving symbol AppModule in /Users/svlad/dev/platform/example-app/app/app.module.ts
```

**Comment by brandonroberts** [comment]
@sharikovvladislav it's a known issue with the CLI. If you save any file and let it recompile it will complete successfully

**Comment by sharikovvladislav** [comment]
What about warnings? Does they matter?

**Comment by sharikovvladislav** [comment]
Ok, if I recompile everything is ok. Lets assume it is not problem. Now I don't have error. But when I open `localhost:4200` I have this: 
```
core.es5.js:1020 ERROR Error: Uncaught (in promise): Error: Cannot find module 'app/books/books.module'.
Error: Cannot find module 'app/books/books.module'.
    at webpackEmptyContext (example-app async:2)
    at SystemJsNgModuleLoader.webpackJsonp.../../../core/@angular/core.es5.js.SystemJsNgModuleLoader.loadAndCompile (core.es5.js:5644)
    at SystemJsNgModuleLoader.webpackJsonp.../../../core/@angular/core.es5.js.SystemJsNgModuleLoader.load (core.es5.js:5632)
    at RouterConfigLoader.webpackJsonp.../../../router/@angular/router.es5.js.RouterConfigLoader.loadModuleFactory (router.es5.js:3417)
    at RouterConfigLoader.webpackJsonp.../../../router/@angular/router.es5.js.RouterConfigLoader.load (router.es5.js:3401)
    at MergeMapSubscriber.project (router.es5.js:1569)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/operator/mergeMap.js.MergeMapSubscriber._tryNext (mergeMap.js:120)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/operator/mergeMap.js.MergeMapSubscriber._next (mergeMap.js:110)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/Subscriber.js.Subscriber.next (Subscriber.js:89)
    at ScalarObservable.webpackJsonp.../../../../rxjs/observable/ScalarObservable.js.ScalarObservable._subscribe (ScalarObservable.js:49)
    at webpackEmptyContext (example-app async:2)
    at SystemJsNgModuleLoader.webpackJsonp.../../../core/@angular/core.es5.js.SystemJsNgModuleLoader.loadAndCompile (core.es5.js:5644)
    at SystemJsNgModuleLoader.webpackJsonp.../../../core/@angular/core.es5.js.SystemJsNgModuleLoader.load (core.es5.js:5632)
    at RouterConfigLoader.webpackJsonp.../../../router/@angular/router.es5.js.RouterConfigLoader.loadModuleFactory (router.es5.js:3417)
    at RouterConfigLoader.webpackJsonp.../../../router/@angular/router.es5.js.RouterConfigLoader.load (router.es5.js:3401)
    at MergeMapSubscriber.project (router.es5.js:1569)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/operator/mergeMap.js.MergeMapSubscriber._tryNext (mergeMap.js:120)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/operator/mergeMap.js.MergeMapSubscriber._next (mergeMap.js:110)
    at MergeMapSubscriber.webpackJsonp.../../../../rxjs/Subscriber.js.Subscriber.next (Subscriber.js:89)
    at ScalarObservable.webpackJsonp.../../../../rxjs/observable/ScalarObservable.js.ScalarObservable._subscribe (ScalarObservable.js:49)
    at resolvePromise (zone.js:770)
    at resolvePromise (zone.js:741)
    at zone.js:818
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:424)
    at Object.onInvokeTask (core.es5.js:3881)
    at ZoneDelegate.webpackJsonp.../../../../zone.js/dist/zone.js.ZoneDelegate.invokeTask (zone.js:423)
    at Zone.webpackJsonp.../../../../zone.js/dist/zone.js.Zone.runTask (zone.js:191)
    at drainMicroTaskQueue (zone.js:584)
    at <anonymous>
```

Planned?

**Comment by brandonroberts** [comment]
@sharikovvladislav no 😄 I saw that issue before, but not consistently. Try changing `app/books/books.module` to `./books/books.module` and save and see if that works

**Comment by sharikovvladislav** [comment]
@brandonroberts That does not work. Same result. A bit another message (because of path changed): `Cannot find module './books/books.module'.`

**Comment by brandonroberts** [comment]
Ok. I'll take a look into it

**Comment by sharikovvladislav** [comment]
Not that I run `ng serve` in root of platform repo

**Comment by sharikovvladislav** [comment]
@brandonroberts I don't know how it is possible, but now everything works fine

**Pull request merged**